File: app/src/test/java/suncor/com/android/ui/main/wallet/cards/CardsViewModelTest.java
package suncor.com.android.ui.main.wallet.cards;

import com.google.gson.Gson;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.Mockito;

import java.util.ArrayList;
import java.util.Arrays;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.MutableLiveData;
import suncor.com.android.data.cards.CardsRepository;
import suncor.com.android.mfp.SessionManager;
import suncor.com.android.model.Resource;
import suncor.com.android.model.account.Profile;
import suncor.com.android.model.cards.CardDetail;
import suncor.com.android.ui.main.wallet.cards.list.CardsViewModel;

public class CardsViewModelTest {

    private CardsViewModel viewModel;
    private CardDetail[] mockCardsList;
    private CardsRepository repository = Mockito.mock(CardsRepository.class);
    private SessionManager sessionManager = Mockito.mock(SessionManager.class);

    @Rule
    public TestRule rule = new InstantTaskExecutorRule();
    private MutableLiveData<Resource<ArrayList<CardDetail>>> cardsLiveData;

    @Before
    public void init() {
        Profile profile = new Profile();
        profile.setPetroPointsNumber("706988500050292051");
        profile.setPointsBalance(2500);
        Mockito.when(sessionManager.getProfile()).thenReturn(profile);
        viewModel = new CardsViewModel(repository, sessionManager);
        //Add a dummy observer to viewState
        viewModel.viewState.observeForever(state -> {
        });
        mockCardsList = new Gson().fromJson(responseJson, CardDetail[].class);

        cardsLiveData = new MutableLiveData<>();
        cardsLiveData.observeForever(arrayListResource -> {
            //A dummy observer
        });
    }

    @Test
    public void test_user_has_only_ppts() {
        ArrayList<CardDetail> cardsList = new ArrayList<>();
        cardsList.add(mockCardsList[0]);
        cardsLiveData.setValue(Resource.success(cardsList));

        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();

        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());
        Assert.assertNotNull(viewModel.getPetroCanadaCards().getValue());
        Assert.assertTrue(viewModel.getPetroCanadaCards().getValue().isEmpty());
        Assert.assertTrue(viewModel.getPartnerCards().getValue().isEmpty());
    }

    @Test
    public void test_user_has_only_petro_canada_cards() {
        ArrayList<CardDetail> cardsList = new ArrayList<>();
        cardsList.add(mockCardsList[0]);
        cardsList.add(mockCardsList[1]);
        cardsList.add(mockCardsList[2]);
        cardsLiveData.setValue(Resource.success(cardsList));

        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();

        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());
        Assert.assertNotNull(viewModel.getPetroCanadaCards().getValue());
        Assert.assertFalse(viewModel.getPetroCanadaCards().getValue().isEmpty());
        Assert.assertTrue(viewModel.getPartnerCards().getValue().isEmpty());
    }

    @Test
    public void test_user_has_only_partner_cards() {

        ArrayList<CardDetail> cardsList = new ArrayList<>();
        cardsList.add(mockCardsList[6]);
        cardsList.add(mockCardsList[7]);
        cardsLiveData.setValue(Resource.success(cardsList));

        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();

        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());
        Assert.assertNotNull(viewModel.getPetroCanadaCards().getValue());
        Assert.assertTrue(viewModel.getPetroCanadaCards().getValue().isEmpty());
        Assert.assertFalse(viewModel.getPartnerCards().getValue().isEmpty());
    }

    @Test
    public void test_user_refresh_success() {
        ArrayList<CardDetail> cardsList = new ArrayList<>(Arrays.asList(mockCardsList));
        cardsLiveData.setValue(Resource.success(cardsList));
        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();
        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());

        viewModel.refreshBalance();
        cardsLiveData.setValue(Resource.loading());
        Assert.assertEquals(CardsViewModel.ViewState.REFRESHING, viewModel.viewState.getValue());

        cardsLiveData.setValue(Resource.success(cardsList));
        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());
        Assert.assertNotNull(viewModel.getPetroCanadaCards().getValue());
        Assert.assertFalse(viewModel.getPetroCanadaCards().getValue().isEmpty());
        Assert.assertFalse(viewModel.getPartnerCards().getValue().isEmpty());
    }

    @Test
    public void test_user_refresh_balance_failed() {
        ArrayList<CardDetail> cardsList = new ArrayList<>(Arrays.asList(mockCardsList));
        cardsLiveData.setValue(Resource.success(cardsList));
        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();
        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());

        viewModel.refreshBalance();
        cardsLiveData.setValue(Resource.loading());
        Assert.assertEquals(CardsViewModel.ViewState.REFRESHING, viewModel.viewState.getValue());

        cardsLiveData.setValue(Resource.error("error", cardsList));
        Assert.assertEquals(CardsViewModel.ViewState.BALANCE_FAILED, viewModel.viewState.getValue());
    }

    @Test
    public void test_user_refresh_failed() {
        ArrayList<CardDetail> cardsList = new ArrayList<>(Arrays.asList(mockCardsList));
        cardsLiveData.setValue(Resource.success(cardsList));
        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();
        Assert.assertEquals(CardsViewModel.ViewState.SUCCESS, viewModel.viewState.getValue());

        viewModel.refreshBalance();
        cardsLiveData.setValue(Resource.loading());
        Assert.assertEquals(CardsViewModel.ViewState.REFRESHING, viewModel.viewState.getValue());

        cardsLiveData.setValue(Resource.error("error"));
        Assert.assertEquals(CardsViewModel.ViewState.FAILED, viewModel.viewState.getValue());
    }

    @Test
    public void test_retrieve_cards_failed() {
        cardsLiveData.setValue(Resource.error("error"));
        Mockito.when(repository.getCards(false)).thenReturn(cardsLiveData);

        viewModel.onAttached();
        Assert.assertEquals(CardsViewModel.ViewState.FAILED, viewModel.viewState.getValue());
        Assert.assertEquals(sessionManager.getProfile().getPetroPointsNumber(), viewModel.getPetroPointsCard().getValue().getCardNumber());
        Assert.assertEquals(sessionManager.getProfile().getPointsBalance(), viewModel.getPetroPointsCard().getValue().getBalance());
        Assert.assertTrue(viewModel.getPetroCanadaCards().getValue() == null || viewModel.getPetroCanadaCards().getValue().isEmpty());
        Assert.assertTrue(viewModel.getPartnerCards().getValue() == null || viewModel.getPartnerCards().getValue().isEmpty());
    }

    private String responseJson = "[{\n" +
            "                        \"cardType\": \"PPTS\",\n" +
            "                        \"cardNumber\": \"706988500050292051\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"11LawF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"pointsBalance\": 12345\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"FSR\",\n" +
            "                        \"cardNumber\": \"706988500050292051\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"25LawF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"litresRemaining\": 12,\n" +
            "                        \"cpl\": 0.05\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"FSR\",\n" +
            "                        \"cardNumber\": \"706988500050292052\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"25LawF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"litresRemaining\": 12,\n" +
            "                        \"cpl\": 0.1\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"PPC\",\n" +
            "                        \"cardNumber\": \"706999700401020280\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"56PawF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"litresRemaining\": 12,\n" +
            "                        \"cpl\": 0.10\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"WAG\",\n" +
            "                        \"cardNumber\": \"706981001000120561\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"11LawF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"unitsRemaining\": 12\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"SP\",\n" +
            "                        \"cardNumber\": \"706981003009633263\",\n" +
            "                        \"cardNumberEncrypted\": \"25AbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"25LawF1v1er4s1D4s0kJEA==\",\n" +
            "                        \"daysRemaining\": 12\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"MORE\",\n" +
            "                        \"cardNumber\": \"60476748007905331\",\n" +
            "                        \"cardNumberEncrypted\": \"198GbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"25LawF1v1er4s1D4s0kJEA==\"\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"CAA\",\n" +
            "                        \"cardNumber\": \"6202909505788005\",\n" +
            "                        \"cardNumberEncrypted\": \"94iIbCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"25LawF1v1er4s1D4s0kJEA==\"\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"HBC\",\n" +
            "                        \"cardNumber\": \"600294472306670\",\n" +
            "                        \"cardNumberEncrypted\": \"33oPCF1v1Or4s1D4s0kJEA==\",\n" +
            "                        \"serviceId\": \"24HuwF1v1er4s1D4s0kJEA==\"\n" +
            "                    },\n" +
            "                    {\n" +
            "                        \"cardType\": \"RBC\"\n" +
            "                    }]";
}

File: app/src/test/java/suncor/com/android/ui/main/wallet/cards/CardsViewModelTest.kt
package suncor.com.android.ui.main.wallet.cards

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.gson.Gson
import org.junit.Assert
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner
import suncor.com.android.data.cards.CardsRepository
import suncor.com.android.mfp.SessionManager
import suncor.com.android.model.Resource
import suncor.com.android.model.account.Profile
import suncor.com.android.model.cards.CardDetail
import suncor.com.android.model.cards.CardType
import suncor.com.android.model.redeem.response.Card
import suncor.com.android.ui.main.wallet.cards.CardsLoadType
import suncor.com.android.ui.main.wallet.cards.list.CardsViewModel
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class CardsViewModelTest {

    private var viewModel: CardsViewModel? = null
    private lateinit var mockCardsList: Array<CardDetail>
    private val repository = Mockito.mock(CardsRepository::class.java)
    private val sessionManager = Mockito.mock(
        SessionManager::class.java
    )
    var profile = Profile()
    private var cardsLiveData: MutableLiveData<Resource<ArrayList<CardDetail>>>? = null

    @get:Rule
    var rule: TestRule = InstantTaskExecutorRule()


    @Before
    fun init() {
        viewModel = CardsViewModel(repository, sessionManager)
        profile.petroPointsNumber = "706988500050292051"
        profile.pointsBalance = 10000

        //Add a dummy observer to viewState
        viewModel!!.viewState.observeForever { state: CardsViewModel.ViewState? -> }
        mockCardsList = Gson().fromJson(responseJson, Array<CardDetail>::class.java)
        cardsLiveData = MutableLiveData()
        cardsLiveData!!.observeForever { arrayListResource: Resource<ArrayList<CardDetail>>? -> }
    }


    @Test
    fun test_profile_availabl_for_load() {
        Assert.assertNotNull(profile.petroPointsNumber)
        Assert.assertNotNull(profile.pointsBalance)

    }

    @Test
    fun test_zero_cardBalance() {
        var isBalanceZero = true
        mockCardsList = Gson().fromJson(responseJson, Array<CardDetail>::class.java)
        val card: CardDetail = mockCardsList[0]
        val apiResponse = MutableLiveData<Resource<CardDetail>>()
        apiResponse.postValue(Resource.success(card))

        val dummyObserver =
            androidx.lifecycle.Observer { event: Resource<CardDetail> -> }
        apiResponse.observeForever(dummyObserver)

        if (card.balance > 0) {
            isBalanceZero = false
            Assert.assertFalse(isBalanceZero)
        } else {
            isBalanceZero = true
            Assert.assertTrue(isBalanceZero)
        }
    }

    @Test
    fun test_user_has_only_ppts() {
        var loadType = CardType.PPTS
        mockCardsList = Gson().fromJson(responseJson, Array<CardDetail>::class.java)
        val card: CardDetail = mockCardsList[0]
        Assert.assertNotNull(card)
        Assert.assertEquals(loadType, card.cardType)
    }

    @Test
    fun test_user_has_only_petro_canada_cards() {
        val cardsList = ArrayList<CardDetail>()
        cardsList.add(mockCardsList[4])
        cardsList.add(mockCardsList[5])
        for (card in cardsList) {
            if (card != null) Assert.assertNotNull(card)
            if (isPetroCanadaCard(card.cardType)) {
                Assert.assertTrue("Is petro canada card", true)
            } else {
                Assert.assertFalse("Is petro canada card", false)
            }
        }
    }

    @Test
    fun test_user_has_only_partner_cards() {
        val cardsList = ArrayList<CardDetail>()
        cardsList.add(mockCardsList[6])
        cardsList.add(mockCardsList[7])
        for (card in cardsList) {
            if (card != null) Assert.assertNotNull(card)
            if (isPartnerCard(card.cardType)) {
                Assert.assertTrue("Is Partner Card", true)
            } else {
                Assert.assertFalse("Not a Partner Card", false)
            }
        }
    }

    @Test
    fun test_getTimeUpdate() {
       viewModel?.dateOfUpdate
    }


    private val responseJson = """[{
                        "cardType": "PPTS",
                        "cardNumber": "706988500050292051",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "11LawF1v1Or4s1D4s0kJEA==",
                        "pointsBalance": 12345
                    },
                    {
                        "cardType": "FSR",
                        "cardNumber": "706988500050292051",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "25LawF1v1Or4s1D4s0kJEA==",
                        "litresRemaining": 12,
                        "cpl": 0.05
                    },
                    {
                        "cardType": "FSR",
                        "cardNumber": "706988500050292052",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "25LawF1v1Or4s1D4s0kJEA==",
                        "litresRemaining": 12,
                        "cpl": 0.1
                    },
                    {
                        "cardType": "PPC",
                        "cardNumber": "706999700401020280",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "56PawF1v1Or4s1D4s0kJEA==",
                        "litresRemaining": 12,
                        "cpl": 0.10
                    },
                    {
                        "cardType": "WAG",
                        "cardNumber": "706981001000120561",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "11LawF1v1Or4s1D4s0kJEA==",
                        "unitsRemaining": 12
                    },
                    {
                        "cardType": "SP",
                        "cardNumber": "706981003009633263",
                        "cardNumberEncrypted": "25AbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "25LawF1v1er4s1D4s0kJEA==",
                        "daysRemaining": 12
                    },
                    {
                        "cardType": "MORE",
                        "cardNumber": "60476748007905331",
                        "cardNumberEncrypted": "198GbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "25LawF1v1er4s1D4s0kJEA=="
                    },
                    {
                        "cardType": "CAA",
                        "cardNumber": "6202909505788005",
                        "cardNumberEncrypted": "94iIbCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "25LawF1v1er4s1D4s0kJEA=="
                    },
                    {
                        "cardType": "HBC",
                        "cardNumber": "600294472306670",
                        "cardNumberEncrypted": "33oPCF1v1Or4s1D4s0kJEA==",
                        "serviceId": "24HuwF1v1er4s1D4s0kJEA=="
                    },
                    {
                        "cardType": "RBC"
                    }]"""

    fun isPetroCanadaCard(cardType: CardType): Boolean {
        return if (cardType.equals(CardType.WAG) || cardType.equals(CardType.SP)) {
            true
        } else {
            false
        }
    }
    fun isPartnerCard(cardType: CardType): Boolean {
        return if (cardType.equals(CardType.RBC) || cardType.equals(CardType.HBC)) {
            true
        } else {
            false
        }
    }
}